### ISSUE TITLE
Example Hang issue

### ISSUE THREAD

**Issue by np2001** [davisking/dlib#1609]
## Expected Behavior

I'm running dnn_introduction_ex example on MNIST database using dlib 19.16 compiled with CMAKE and CUDA 10.0 support. 

## Current Behavior
Example starts correctly and training runs succesfully, but after few epochs it just randomly hangs and stops to output any progress text to console. 

Epoch: 1     learning rate: 0.01  average loss: 0.294481     steps without apparent progress: 8
Epoch: 2     learning rate: 0.01  average loss: 0.0830322    steps without apparent progress: 15
Epoch: 3     learning rate: 0.01  average loss: 0.058174     steps without apparent progress: 5
Epoch: 4     learning rate: 0.01  average loss: 0.0475068    steps without apparent progress: 9
Epoch: 5     learning rate: 0.01  average loss: 0.0398031    steps without apparent progress: 357
Epoch: 6     learning rate: 0.01  average loss: 0.0343465    steps without apparent progress: 345
Epoch: 7     learning rate: 0.01  average loss: 0.0305033    steps without apparent progress: 10
Epoch: 8     learning rate: 0.01  average loss: 0.0274168    steps without apparent progress: 355
Saved state to mnist_sync
Epoch: 9     learning rate: 0.01  average loss: 0.0240439    steps without apparent progress: 359
Epoch: 10    learning rate: 0.01  average loss: 0.0212106    steps without apparent progress: 11
- - - - - - - - - - - - - - - - - - 
Epoch: 27    learning rate: 0.01  average loss: 0.00589464   steps without apparent progress: 1043
Epoch: 28    learning rate: 0.01  average loss: 0.00564767   steps without apparent progress: 1387
Epoch: 29    learning rate: 0.01  average loss: 0.00553218   steps without apparent progress: 1316
Saved state to mnist_sync
Epoch: 30    learning rate: 0.01  average loss: 0.00536734   steps without apparent progress: 1448
Epoch: 31    learning rate: 0.01  average loss: 0.00529126   steps without apparent progress: 1498

Pausing the debugger points to call of CHECK_CUDA(cudaStreamSynchronize(0)) method of gpu_data::async_copy_to_device() method (source file gpu_data.cpp Line 130). Also Visual Studio Output window shows some file_not_found exceptions during example run, that are not propagated to example's catch block:

Exception thrown at 0x00007FFFC1864048 in dnn_introduction_ex.exe: Microsoft C++ exception: dlib::file::file_not_found at memory location 0x000000E970AFE1B0.

Cant't definitly say if it relates to the initial problem.

## Steps to Reproduce
Just normal run of an example, preferably on Release configurations. Debug configuration seems to finish correctly, or maybe just has less chance to hang.

* **Version**: Dlib 19.16
* **Where did you get dlib**: Release from dlib.net
* **Platform**: Windows 10 64 bit
* **Compiler**: Visual Studio 2017 (15.9.3), Cuda 10.0 with CuDNN 7.4
* **Hardware** Intel Core i5-4670, 16 GB Ram, GeForce 1050 Ti

**Comment by davisking** [comment]
This seems to be a bug in cuda. See https://github.com/davisking/dlib/pull/1514. If you try the latest dlib from github does it go away?

**Comment by np2001** [comment]
Well, this problem seems to be worked-around in the latest dlib version. And, though, busy-looping is not looks like an efficient solution, cudaStreamSynchronize (while it not hangs) consumes about the same CPU resources.

**Comment by davisking** [comment]
A lot of cuda calls busy-wait internally anyway, so that's not a big deal.  More troubling is that the cuda bug seems to be not totally avoided. See https://github.com/davisking/dlib/pull/1596.
